File: frontend/src/incidents.ts
import { apiBaseUrl } from './lib'

export type IncidentEvent = {
  id: number
  event_type: string
  event_at: string
  event_label: string | null
  metadata: Record<string, unknown>
}

export type Incident = {
  id: number
  external_id: string
  dataset_name: string
  source: string | null
  initiated_by: string | null
  reported_at: string
  closed_at: string | null
  status: string
  raw_payload: Record<string, unknown>
  events: IncidentEvent[]
}

export type IncidentLocation = {
  id: number
  canonical_address: string
  boro: string
  house_num: string | null
  street_name: string
  from_street: string | null
  to_street: string | null
  spec_loc: string | null
  location_key: string
}

export type GeoJsonLineString = {
  type: 'LineString'
  coordinates: [number, number][]
}

export type GeoJsonMultiLineString = {
  type: 'MultiLineString'
  coordinates: [number, number][][]
}

export type IncidentMapData = {
  normalized_address: string
  location: IncidentLocation
  geometry: GeoJsonLineString | GeoJsonMultiLineString
  bbox: [number, number, number, number] | null
  center: [number, number] | null
}

export type AddressIncidentLookupResponse = {
  address: string
  normalized_address: string
  location: IncidentLocation
  map: IncidentMapData | null
  incidents: Incident[]
  incident_count: number
  event_count: number
}

export type LiabilitySignal = 'likely_liable' | 'likely_not_liable'
export type CaseStrength = 'strong' | 'maybe' | 'weak'

export type LiabilityAnalysisResponse = {
  address: string
  client_incident_date: string
  liability_signal: LiabilitySignal
  case_strength: CaseStrength
  best_matching_incident_id: string | null
  best_matching_days_open: number | null
  analysis_summary: string
  disclaimer: string
}

export class IncidentLookupError extends Error {
  status: number

  constructor(message: string, status: number) {
    super(message)
    this.name = 'IncidentLookupError'
    this.status = status
  }
}

export async function fetchIncidentsByAddress(
  address: string,
  accessToken: string,
): Promise<AddressIncidentLookupResponse> {
  const response = await fetch(`${apiBaseUrl}/api/incidents/by-address`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify({ address }),
  })

  if (!response.ok) {
    let detail = 'Something went wrong. Please try again.'

    try {
      const payload = (await response.json()) as { detail?: string }
      if (payload.detail) {
        detail = payload.detail
      }
    } catch {
      // Ignore non-JSON error bodies and fall back to generic messaging.
    }

    throw new IncidentLookupError(detail, response.status)
  }

  return (await response.json()) as AddressIncidentLookupResponse
}

export async function fetchLiabilityAnalysis(
  address: string,
  clientIncidentDate: string,
  accessToken: string,
): Promise<LiabilityAnalysisResponse> {
  const response = await fetch(`${apiBaseUrl}/api/incidents/liability-analysis`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify({
      address,
      client_incident_date: clientIncidentDate,
    }),
  })

  if (!response.ok) {
    let detail = 'Something went wrong. Please try again.'

    try {
      const payload = (await response.json()) as { detail?: string }
      if (payload.detail) {
        detail = payload.detail
      }
    } catch {
      // Ignore non-JSON error bodies and fall back to generic messaging.
    }

    throw new IncidentLookupError(detail, response.status)
  }

  return (await response.json()) as LiabilityAnalysisResponse
}

File: backend/app/models/incidents.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class IncidentLookupRequest(BaseModel):
    address: str | None = None


class IncidentEventResponse(BaseModel):
    id: int
    event_type: str
    event_at: datetime
    event_label: str | None
    metadata: dict[str, Any]


class IncidentResponse(BaseModel):
    id: int
    external_id: str
    dataset_name: str
    source: str | None
    initiated_by: str | None
    reported_at: date
    closed_at: date | None
    status: str
    raw_payload: dict[str, Any]
    events: list[IncidentEventResponse]


class LocationResponse(BaseModel):
    id: int
    canonical_address: str
    boro: str
    house_num: str | None
    street_name: str
    from_street: str | None
    to_street: str | None
    spec_loc: str | None
    location_key: str


class GeoJsonLineGeometry(BaseModel):
    type: str
    coordinates: list[Any]


class IncidentMapData(BaseModel):
    normalized_address: str
    location: LocationResponse
    geometry: GeoJsonLineGeometry
    bbox: list[float] | None = None
    center: list[float] | None = None


class AddressIncidentLookupResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    address: str
    normalized_address: str
    location: LocationResponse
    map: IncidentMapData | None = None
    incidents: list[IncidentResponse]
    incident_count: int
    event_count: int


class DatabaseDebugResponse(BaseModel):
    current_database: str
    location_count: int
    incident_count: int
    event_count: int
    sample_location: LocationResponse | None


class AddressLookupDebugResponse(BaseModel):
    input_address: str
    normalized_address: str
    match_count: int
    matches: list[LocationResponse]


class LiabilityAnalysisRequest(BaseModel):
    address: str | None = None
    client_incident_date: date | None = None


class LiabilityIncidentSummary(BaseModel):
    external_id: str
    reported_at: date
    closed_at: date | None
    status: str
    days_open_as_of_client_incident: int | None
    was_open_on_client_incident_date: bool
    qualifies_for_notice_window: bool


class LiabilityAnalysisResponse(BaseModel):
    address: str
    client_incident_date: date
    liability_signal: str
    case_strength: str
    best_matching_incident_id: str | None
    best_matching_days_open: int | None
    analysis_summary: str
    disclaimer: str


class LiabilityPromptPayload(BaseModel):
    address: str
    client_incident_date: date
    liability_signal: str
    case_strength: str
    best_matching_incident: LiabilityIncidentSummary | None
    incident_summaries: list[LiabilityIncidentSummary]
    additional_supporting_incident_count: int = Field(default=0, ge=0)
